### PR TITLE
issue-37 - Optimization for NodeActivation job

### PR DESCRIPTION
**Pull request by ivouchak-sc**

### CODE REVIEW

File: grails-app/jobs/NodeActivatorJob.groovy [hunk 1]
```diff
@@ -72,49 +71,58 @@ class NodeActivatorJob {
         try{
             if (log.debugEnabled) {
                 sw?.start('Finding active and running FlowStatuses')
```
Comment (apechenko-sc) [inline grails-app/jobs/NodeActivatorJob.groovy:73]: If you don't have code between lines 73 and 74 which executes some heavy task, then it doesn't make sense to start and stop the stop watch.

File: grails-app/jobs/NodeActivatorJob.groovy [hunk 2]
```diff
@@ -72,49 +71,58 @@ class NodeActivatorJob {
...
+                    processNodeTypes: [
+                            ConstantUtils.NODE_TYPE_ACTIVITY,
+                            ConstantUtils.NODE_TYPE_FORK,
+                            ConstantUtils.NODE_TYPE_ORJOIN,
+                            ConstantUtils.NODE_TYPE_ANDJOIN
+                    ]
+            ]
+
+            List<Long> activeNodesKeys = ProcessNode.executeQuery("""
+
```
Comment (apechenko-sc) [inline grails-app/jobs/NodeActivatorJob.groovy:93]: Is the empty line needed here?

File: grails-app/jobs/NodeActivatorJob.groovy [hunk 3]
```diff
@@ -72,49 +71,58 @@ class NodeActivatorJob {
...
             if (activeNodesKeys) {
                 if (log.debugEnabled) {
                     sw?.start('Loading ProcessNodes by keys')
                 }
 
-                List activeNodes = []
-                activeNodesKeys.each {
-                    activeNodes << ProcessNode.get(it)
-                }
+                List<ProcessNode> activeNodes = ProcessNode.findAllByIdInList(activeNodesKeys)
```
Comment (apechenko-sc) [inline grails-app/jobs/NodeActivatorJob.groovy:118]: Does it make sense to load IDs first and then load objects? Isn't it better to load objects already in line 92?

File: grails-app/services/com/jcatalog/grailsflow/ProcessManagerService.groovy [hunk 4]
```diff
@@ -244,7 +244,7 @@ class ProcessManagerService implements InitializingBean {
      *
      * @return process ID value if process was started successfully
      */
-    public synchronized def startProcess(def processTypeID, def user, def variables) {
+    public def startProcess(def processTypeID, def user, def variables) {
```
Comment (apechenko-sc) [inline grails-app/services/com/jcatalog/grailsflow/ProcessManagerService.groovy:247]: Looks dangerous. Should be tested heavily.

Comment (ivouchak-sc) [reply grails-app/services/com/jcatalog/grailsflow/ProcessManagerService.groovy:247]: Locally works normally: 10 start threads. 
Start process is independent part and no need to synchronize it.

File: grails-app/jobs/NodeActivatorJob.groovy (hunk 2 above)
Comment (ivouchak-sc) [reply grails-app/jobs/NodeActivatorJob.groovy:93]: Only for better readability of HQL.

File: grails-app/jobs/NodeActivatorJob.groovy (hunk 3 above)
Comment (ivouchak-sc) [reply grails-app/jobs/NodeActivatorJob.groovy:118]: The previous approach was to get rid of criteria and use plain SQL. But now we have a limit to getting process nodes per one request and we make to try it.

File: grails-app/jobs/NodeActivatorJob.groovy [hunk 5]
```diff
@@ -71,50 +70,53 @@ class NodeActivatorJob {
...
+            log.info "*** Amount of Nodes to execute ${activeNodes.size()} ***"
 
-                List activeNodes = []
-                activeNodesKeys.each {
-                    activeNodes << ProcessNode.get(it)
-                }
+            def nodesComparator = grailsApplication.config.grailsflow.nodeActivator.comparator
 
+            if (activeNodes) {
                 if (log.debugEnabled) {
```
Comment (apechenko-sc) [inline grails-app/jobs/NodeActivatorJob.groovy:109]: IF block can be removed now.

**Pull request merged**